### PR TITLE
feat: add dotcom option to view all subscriptions ordered by license …

### PR DESCRIPTION
**Pull request by dadlerj**
…expiration date

Fixes https://github.com/sourcegraph/sourcegraph/issues/3017

NOT a priority for 3.4, no need to look at this until next week, just didn't want this to sit for a week after being finished yesterday...

No changelog entry needed due to this only applying on dotcom?

I don't think there's a super easy pattern for removing the decode functionality from license.go entirely, but it is worth looking into it before merging this so we don't have duplicate base64 license key decoding happening (now in both license.go in Golang AND in subscriptions_db.go in SQ). 

Finally, not sure about who the backend code owner should be for subscription stuff going forward, but I assume someone on Core :)

### CODE REVIEW

File: cmd/frontend/graphqlbackend/dotcom.go [hunk 1]
```diff
@@ -135,6 +135,7 @@ type ProductSubscriptionArgs struct {
 type ProductSubscriptionsArgs struct {
 	graphqlutil.ConnectionArgs
 	Account *graphql.ID
+	OrderBy string
```
Comment (felixfbecker) [inline cmd/frontend/graphqlbackend/dotcom.go]: Should this be an enum type? https://sourcegraph.sgdev.org/search?q=r:sourcegraph/sourcegraph%24+orderBy+lang:go+

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go [hunk 2]
```diff
@@ -86,16 +110,40 @@ func (o dbSubscriptionsListOptions) sqlConditions() []*sqlf.Query {
...
+)
+SELECT id, user_id, billing_subscription_id, created_at, archived_at, (
+	CASE WHEN active_key IS NULL THEN NULL ELSE CAST(
+		CONVERT_FROM(
+			DECODE(
+				CONVERT_FROM(
+					DECODE(
+						CONCAT(
+							TRIM(active_key),
+							REPEAT('=', MOD(4 - MOD(CHAR_LENGTH(active_key), 4), 4))),
```
Comment (felixfbecker) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: Why is `MOD` used twice? Any `mod 4` should return a natural number from 0 to 4, so then that number subtracted from 4 should yield a natural number from 0 to 4 too. Why are you taking the mod 4 again after that?

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go [hunk 3]
```diff
@@ -86,16 +110,40 @@ func (o dbSubscriptionsListOptions) sqlConditions() []*sqlf.Query {
...
+func (dbSubscriptions) list(ctx context.Context, conds []*sqlf.Query, orderBy SubscriptionsListOrderBy, limitOffset *db.LimitOffset) ([]*dbSubscription, error) {
 	q := sqlf.Sprintf(`
-SELECT id, user_id, billing_subscription_id, created_at, archived_at FROM product_subscriptions
+WITH active_license_keys as (
+	SELECT product_subscription_id, active_key FROM (
+		SELECT product_subscription_id, FIRST_VALUE(license_key) over (
+			PARTITION by product_subscription_id ORDER BY created_at DESC) AS active_key
+		FROM product_licenses) AS licenses_with_active_keys
+	GROUP BY product_subscription_id, active_key
+)
```
Comment (felixfbecker) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: I have to admit I don't really understand what this does with `WITH`, `FIRST_VALUE` and `PARTITION` (never used those). You sure this is needed?

File: web/src/enterprise/site-admin/dotcom/productSubscriptions/SiteAdminProductSubscriptionsPage.tsx [hunk 4]
```diff
@@ -25,10 +26,23 @@ class FilteredSiteAdminProductSubscriptionConnection extends FilteredConnection<
     Pick<SiteAdminProductSubscriptionNodeProps, 'onDidUpdate'>
 > {}
 
+type subscriptionsDisplays = 'by-created-at' | 'by-expires-at'
```
Comment (felixfbecker) [inline web/src/enterprise/site-admin/dotcom/productSubscriptions/SiteAdminProductSubscriptionsPage.tsx]: PascalCase type names

**Comment by sqs** [comment]
You could consider simplifying this by just reading in ALL licenses and performing the parsing/filtering/sorting in Go instead of duplicating it in SQL. Even with 10,000s of licenses, it's not going to be so slow for the operations you need.

(Totally your call - this is purely for internal consumption anyway, so it's not the end of the world if this breaks or falls out of sync, it's easy to fix if so)

**Comment by felixfbecker** [comment]
I think it’s cleaner to do it in SQL because then it can be easily combined with other filters without having to reimplement those filters in Go too

**Comment by dadlerj** [comment]
Yeah, @sqs the discussion that drove this was all around a general question of how to handle ordering/filtering that happens "offline" (ie, outside of SQL). See https://sourcegraph.slack.com/archives/C07KZF47K/p1557865845074800 for context.

The perf implications are minimal either way, given the internal nature of this particular operation, this is a broader question about coding style that I look to the dev team to direct me on.

If you have any strong opinions, lmk. Curious what Keegan (or whoever reviews the backend code), thinks as well.

File: cmd/frontend/graphqlbackend/dotcom.go (hunk 1 above)
Comment (keegancsmith) [reply cmd/frontend/graphqlbackend/dotcom.go]: Look at other places we have orderby in the graphqlbackend. We just use string (or `*string`). Our go graphql layer doesn't enforce enums.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go [hunk 5]
```diff
@@ -86,16 +110,40 @@ func (o dbSubscriptionsListOptions) sqlConditions() []*sqlf.Query {
...
+							TRIM(active_key),
+							REPEAT('=', MOD(4 - MOD(CHAR_LENGTH(active_key), 4), 4))),
+						'base64'),
+					'UTF-8')::json->>'info',
+				'base64'),
+			'UTF-8')::json->>'e'
+		AS TIMESTAMP WITH TIME ZONE) END
+) as expires_at
+FROM product_subscriptions
+LEFT JOIN active_license_keys ON active_license_keys.product_subscription_id = product_subscriptions.id
```
Comment (keegancsmith) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: Ensure this SQL works on our lowest supported version of postgres.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go (hunk 3 above)
Comment (keegancsmith) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: cc @mrnugget I think he has some SQL skills. But I think the with table essentially is the latest license_key per production_subscription_id. This unfortunately is likely the best way to do that to my knowledge, although you may be able to get rid of one level of selects:

```sql
WITH active_license_keys as (
  SELECT
    product_subscription_id,
    first_value(license_key) OVER (PARTITION BY product_subscription_id ORDER BY created_at DESC) AS active_key
  FROM product_licenses
  GROUP BY product_subscription_id
)
```

no idea if that works though, haven't tested it.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go [hunk 6]
```diff
@@ -86,16 +110,40 @@ func (o dbSubscriptionsListOptions) sqlConditions() []*sqlf.Query {
...
+				CONVERT_FROM(
+					DECODE(
+						CONCAT(
+							TRIM(active_key),
+							REPEAT('=', MOD(4 - MOD(CHAR_LENGTH(active_key), 4), 4))),
+						'base64'),
+					'UTF-8')::json->>'info',
+				'base64'),
+			'UTF-8')::json->>'e'
+		AS TIMESTAMP WITH TIME ZONE) END
```
Comment (keegancsmith) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: I was wondering why you did this gymnastics, but I see you want to order by expires_at. This code makes me nervous to be honest, so I would at least want a unit test for it. 

Alternative ideas:
- Migration using this code which adds an `expires_at` column.
- Migration storing the active_key as a json column, rather than b64.
- Do the order by expires in go code. As it stands we don't have that many license keys.

However, this is only used by sourcegraph.com? Do server pings trigger this code? How often will it get run? As it stands if it is not that often and we have a unit test for this code, I'm happy with using an overly complex SQL statement.

File: enterprise/pkg/license/license.go [hunk 7]
```diff
@@ -87,7 +87,9 @@ func (l Info) encode() ([]byte, error) {
 	if _, err := rand.Read(e.Nonce[:8]); err != nil {
 		return nil, err
 	}
-	return json.Marshal(e)
+	x, err := json.Marshal(e)
+	fmt.Println(x, err)
```
Comment (keegancsmith) [inline enterprise/pkg/license/license.go]: debug output

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go (hunk 2 above)
Comment (dadlerj) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: `mod 4` returns 0–3, not 0–4. Then, `4 - mod 4` returns 1–4. This second `mod` is then necessary to get it back to 0–3!

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go (hunk 3 above)
Comment (dadlerj) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: @felixfbecker each `product_subscription` can have multiple `product_license`s. What I need to do (and what this `WITH` clause does) is, for each `product_subscription_id`, find the most recently created license_key associated with it. If you know a better way to calculate this, let me know, but I use SQL quite a bit (typically BigQuery-flavored though) and this is definitely the most straightforward path that I'm aware of.

The inner `SELECT` here creates a table that has exactly the same number of rows as `product_licenses`, but where each row now has a new column that captures the latest (chronologically) `license_key` associated with that row's `product_subscription_id`, e.g.:

If `product_licenses` contained:

| product_subscription_id | license_key | created_at |
| --- | --- | --- |
| A | 123 | last month |
| A | 456 | today |
| B | 000 | yesterday |
| A | 789 | a year ago |


The inner clause would create:


|product_subscription_id|active_key|
| --- | --- |
|A|456|
|A|456|
|B|000|
|A|456|


The outer `SELECT` then takes all of the duplicate rows in that table and aggregates by uniques, e.g.:

|product_subscription_id|active_key|
| --- | --- |
|A|456|
|B|000|

@keegancsmith unfortunately this has to happen in two steps — it's a quirk of window functions, which, critically, don't behave the same as aggregate functions!

Your command returns:
>ERROR:  column "product_licenses.license_key" must appear in the GROUP BY clause or be used in an aggregate function

And if you try to group on it you get:
>ERROR:  window functions are not allowed in GROUP BY

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go (hunk 6 above)
Comment (dadlerj) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: Yeah, I just don't think doing all of this via SQL is worth it here. I've made the call to remove this and do it all in Go instead...

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go (hunk 3 above)
Comment (dadlerj) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_db.go]: But given the risk of putting business logic into SQL, I'm just going to do this in Go instead of SQL anyway, so no need to go further here 😄. 

There'll never really be a single solution to this problem of sorting in multiple places, anyway. E.g., in this case, the sort that I thought we couldn't do in SQL actually could be done there (even if it was a bit ugly), but in other cases (e.g. for sorting users by "last active date"), we have to query Redis, which really would be impossible...

https://github.com/sourcegraph/sourcegraph/pull/4110#discussion_r285776259

File: web/src/components/time/Timestamp.tsx [hunk 8]
```diff
@@ -41,7 +41,7 @@ export class Timestamp extends React.PureComponent<Props> {
             label = label.replace('about ', '')
         }
         return (
-            <span className={`timestamp${' ' + this.props.className || ''}`} data-tooltip={this.props.date}>
+            <span className={`timestamp ${this.props.className || ''}`} data-tooltip={this.props.date}>
```
Comment (keegancsmith) [inline web/src/components/time/Timestamp.tsx]: why did you change this? Look at the effects on the snapshot below. Seems like the code before was better?

Comment (felixfbecker) [reply web/src/components/time/Timestamp.tsx]: To be able to add a className. Spaces in `className` don't really matter but you can avoid them by using the `classNames()` helper function.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go [hunk 9]
```diff
@@ -473,13 +492,54 @@ func (ProductSubscriptionLicensingResolver) ProductSubscriptions(ctx context.Con
...
+}
+
+func (s sortSubscriptions) Less(i, j int) bool {
+	// Subscriptions with no license should always be last.
+	if s[i].activeLicense == nil {
+		return false
+	}
+	if s[j].activeLicense == nil {
+		return true
+	}
```
Comment (keegancsmith) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go]: you are not handling the case of them both being nil. In that case you need to return false.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go [hunk 10]
```diff
@@ -473,13 +492,54 @@ func (ProductSubscriptionLicensingResolver) ProductSubscriptions(ctx context.Con
 type productSubscriptionConnection struct {
 	opt dbSubscriptionsListOptions
 
+	// special handling for non-SQL ordering option
+	orderByExpiresAt bool
+
 	// cache results because they are used by multiple fields
 	once    sync.Once
-	results []*dbSubscription
+	results sortSubscriptions
```
Comment (keegancsmith) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go]: nit: You don't need to change this. The sorting code should be isolated to the compute function.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go [hunk 11]
```diff
@@ -488,7 +548,28 @@ func (r *productSubscriptionConnection) compute(ctx context.Context) ([]*dbSubsc
...
+
+		r.results = make([]*subscriptionWithActiveLicense, len(dbSubs))
+		for i, s := range dbSubs {
+			r.results[i] = &subscriptionWithActiveLicense{subscription: s}
+		}
+
+		if r.orderByExpiresAt {
+			for i, s := range r.results {
+				ps := &productSubscription{v: s.subscription}
+				r.results[i].activeLicense, r.err = ps.activeDBLicense(ctx)
```
Comment (keegancsmith) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go]: this is not great. I think your funky SQL was better. This will now do a DB query per subscription. If you want to go this route, rather directly use `dbLicenses{}.List` and list everything and pair it up.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go [hunk 12]
```diff
@@ -473,13 +492,54 @@ func (ProductSubscriptionLicensingResolver) ProductSubscriptions(ctx context.Con
...
+	// Subscriptions with no license should always be last.
+	if s[i].activeLicense == nil {
+		return false
+	}
+	if s[j].activeLicense == nil {
+		return true
+	}
+
+	l1 := &productLicense{v: s[i].activeLicense}
+	l1Info, err := l1.Info()
```
Comment (keegancsmith) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go]: what work does this do? If it is CPU or IO intensive it shouldn't be inside of the sort.

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go (hunk 9 above)
Comment (dadlerj) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/subscriptions_graphql.go]: Wouldn't that case satisfy _both_ of the `if`s above, and thus it'll always return on the first one (i.e., it will return `false` correctly)?

**Comment by dadlerj** [comment]
fyI no need to review this as-is, latest push was only for testing saved searches...

File: enterprise/pkg/license/license.go [hunk 13]
```diff
@@ -147,5 +147,6 @@ func ParseSignedKey(text string, publicKey ssh.PublicKey) (info *Info, signature
 	if err := json.Unmarshal(signedKey.EncodedInfo, &info); err != nil {
 		return nil, "", err
 	}
+	fmt.Println("verifying, " + string(signedKey.EncodedInfo))
```
Comment (beyang) [inline enterprise/pkg/license/license.go]: Don't use `fmt.Println`. If you want to log this, then we prefer [`log15.Info`](https://godoc.org/github.com/inconshreveable/log15#Info) as our logging function.

Comment (dadlerj) [reply enterprise/pkg/license/license.go]: Thanks — yeah, that was debug output, this is still just wip until I get tests in...

**Comment by slimsag** [comment]
This PR has been open for a very long time with no progress, closing. Reopen if desired.
